viz/pkg/healthcheck/healthcheck.go

```diff
@@ -54,147 +72,242 @@ func (hc *HealthChecker) RunChecks(observer healthcheck.CheckObserver) bool {
+				Fatal().
+				WithCheck(func(ctx context.Context) (err error) {
+					hc.vizAPIClient, err = client.NewExternalClient(ctx, hc.vizNamespace, hc.KubeAPIClient())
+					return
+				}),
+			*healthcheck.NewChecker("data plane proxy metrics are present in Prometheus").
+				WithHintAnchor("l5d-data-plane-prom").
+				Warning().
+				WithRetryDeadline(hc.RetryDeadline).
+				WithCheck(func(ctx context.Context) (err error) {
+					pods, err := hc.getDataPlanePodsFromVizAPI(ctx)
+					if err != nil {
+						return err
+					}
+
+					// TODO: Check if prometheus is present
+
+					return validateDataPlanePodReporting(pods)
+				}),
+			*healthcheck.NewChecker("viz extension self-check").
+				WithHintAnchor("l5d-api-control-api").
+				Fatal().
+				// to avoid confusing users with a prometheus readiness error, we only show
+				// "waiting for check to complete" while things converge. If after the timeout
+				// it still hasn't converged, we show the real error (a 503 usually).
+				WithRetryDeadline(hc.RetryDeadline).
+				WithCheckRPC(func(ctx context.Context) (*healthcheckPb.SelfCheckResponse, error) {
+					return hc.vizAPIClient.SelfCheck(ctx, &healthcheckPb.SelfCheckRequest{})
+				}),
+		}, true),
+
+		*healthcheck.NewCategory(LinkerdVizExtensionDataPlaneCheck, []healthcheck.Checker{
+			*healthcheck.NewChecker("data plane namespace exists").
+				WithHintAnchor("l5d-data-plane-exists").
+				Fatal().
+				WithCheck(func(ctx context.Context) error {
+					if hc.DataPlaneNamespace == "" {
+						// when checking proxies in all namespaces, this check is a no-op
+						return nil
+					}
+					return hc.CheckNamespace(ctx, hc.DataPlaneNamespace, true)
+				}),
+			*healthcheck.NewChecker("data-plane pods have the correct tap configuration").
+				WithHintAnchor("l5d-viz-data-plane-tap").
+				Fatal().
+				WithCheck(func(ctx context.Context) error {
+					pods, err := hc.GetDataPlanePods(ctx)
+					if err != nil {
+						return err
+					}
+
+					// Check environment specific configuration based on labels
+					// Do not warn for the viz extension or core cp pods
+					return hc.checkForTapConfiguration(ctx, pods)
+				}),
+		}, true),
+	}
+}
+
+func (hc *HealthChecker) getDataPlanePodsFromVizAPI(ctx context.Context) ([]*pb.Pod, error) {
+
+	req := &pb.ListPodsRequest{}
+	if hc.DataPlaneNamespace != "" {
+		req.Selector = &pb.ResourceSelection{
+			Resource: &pb.Resource{
+				Namespace: hc.DataPlaneNamespace,
+			},
+		}
+	}
+
+	resp, err := hc.VizAPIClient().ListPods(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	pods := make([]*pb.Pod, 0)
+	for _, pod := range resp.GetPods() {
+		if pod.ControllerNamespace == hc.ControlPlaneNamespace {
+			pods = append(pods, pod)
+		}
+	}
+
+	return pods, nil
+}
+
+// checkForTapConfiguration checks if the tap env variable is present in the linkerd proxy
+// only for the pods with tap enabled
+func (hc *HealthChecker) checkForTapConfiguration(ctx context.Context, pods []corev1.Pod) error {
```

Comment:
Note that when we implement https://github.com/linkerd/linkerd2/issues/5544, we will add an annotation to pods which have been tap-injected.  This will simplify this method since we can just look for that annotation instead of looking at the environment variables.

Reply:
Yep.